orderer/common/cluster/block_fetcher.go

```diff
@@ -0,0 +1,503 @@
+/*
+Copyright IBM Corp. 2022 All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package cluster
+
+import (
+	"context"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/hyperledger/fabric-protos-go/common"
+	"github.com/hyperledger/fabric-protos-go/orderer"
+	ab "github.com/hyperledger/fabric-protos-go/orderer"
+	"github.com/hyperledger/fabric/common/flogging"
+	"github.com/hyperledger/fabric/common/util"
+	"github.com/hyperledger/fabric/internal/pkg/identity"
+	"github.com/hyperledger/fabric/protoutil"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+)
+
+type BlockSourceOps int
+
+const (
+	CURRENTSOURCE BlockSourceOps = iota
+	SHUFFLESOURCE
+	SUSPECTSOURCE
+)
+
+//go:generate mockery -dir . -name BlockSource -case underscore -output mocks/
+
+type BlockSource interface {
+	PullBlock(seq uint64) *common.Block
+	UpdateEndpoints(endpoints []EndpointCriteria)
+	Close()
+}
+
+type attestationErrorAndResponse struct {
+	err  error
+	resp *orderer.BlockAttestationResponse
+}
+
+type attestationConnectionInfo struct {
+	stream *impatientAttestationStream
+	conn   *grpc.ClientConn
+}
+
+//go:generate mockery -dir . -name AttestationSource -case underscore -output mocks/
+type AttestationSource interface {
+	PullAttestations(seq uint64) []*orderer.BlockAttestation
+	Close()
+}
+
+type suspectSet struct {
+	cache     map[string]uint
```

Comment:
you should evict them by their order of addition. Why do you need a map? Just have a slice... 
```
var suspects []string

// add suspect to slice:
for len(suspects) >= max {
      suspects = suspects[1:]
}
suspects[len(suspects)-1] = newEndpoint
```

There is no point in efficiently looking up if an endpoint is a suspect using a map, the endpoints are small and there is only a few of them, so a for loop over the suspects is good enough. No need to make the code more complex than that :-)